doc/proposals/metering-operator-metrics.md

```diff
@@ -0,0 +1,99 @@
+## Auto register operator specific metrics as part of operator-metering
+
+### Motivation and goal
+
+We want to be able to generate the metering reports based on the operator specific Prometheus metrics. In order to be able to do that, operators must be instrumented to expose those metrics, and the operator-sdk should make this as easy as possible. The goal is to have the metering happen based on the usage of each individual operator. Metrics will be based on objects managed by the particular operator.
+
+### Overview of the metrics
+
+To follow both the Prometheus instrumentation [best practices](https://prometheus.io/docs/practices/naming/) as well as the official Kubernetes instrumentation [guide](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md), the metrics will have the following format:
+
+```
+crd_kind_info{namespace="namespace",crdkind="instance-name"} 1
+```
+
+example metric for the memchached-operator would look like this:
+
+```
+memcached_info{namespace="default",memcached="example-memcached"} 1
+```
+
+### Existing solutions
+
+Either solutions makes use of Kubernetes list/watch to populate a Prometheus metrics registry, only that kube-state-metrics solution implements its own registry for performance reasons and the client_golang approach makes use of the canonical metrics registries.
+
+#### kube-state-metrics based
+
+[kube-state-metrics](https://github.com/kubernetes/kube-state-metrics#overview) can be used because it solves exactly the same problem but it does it for upstream known resources. The operator-sdk can re-use its functionality do perform the same thing with custom resources.
+
+```go
+// NewCollectors returns a collection of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+func NewCollectors(
+client *Client,
+	namespaces []string,
+	api string,
+	kind string,
+	metricsGenerator func(obj interface{}) []*metrics.Metric) (collectors []*kcoll.Collector)
+```
+
+```go
+// ServeMetrics takes in the collectors that were created and port number on which the metrics will be served.
+func ServeMetrics(collectors []*kcoll.Collector, portNumber int) {
+
+```
+
+Note: Due to taking advantage of kube-state-metrics functions and interfaces we cannot use the prometheus/client_golang and we need to register it in the same way as kube-state-metrics does, and expose the `/metrics` and serve it on a port (port `:8389/metrics` for example). For that we will need to also create a Service object or rather update the current Service object.
+
+#### Using Prometheus client_golang
+
+The advantage of using the Prometheus native [client-golang](https://github.com/prometheus/client_golang) is we can use the already existing metrics registry and therefore not need to serve another http server for the metrics as we would need to in the first solution. 
```

Comment:
You might want two different ports for this anyways?!

Kube-state-metrics exposes one port about internal metrics (e.g. count of http requests to its endpoints) and one port of Kubernetes object metrics (its soul purpose). Maybe it makes sense to split the two for an operator as well?

Reply:
Yes, I noticed that in kube-state-metrics, but didn't know the reasoning behind the split?

If I understand correctly for Prometheus it doesn't make much of a difference, the only reason is we would need to serve the metrics twice, and if we can avoid that it would make things easier when exposing the metrics ports later on. But for me personally I am fine with either solutions.